query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/QueryRequestToPinotSQLConverter.java

```diff
@@ -154,40 +154,37 @@ private String convertFilter2String(
           break;
         case CONTAINS_KEY:
           LiteralConstant[] kvp = convertExpressionToMapLiterals(filter.getRhs());
-          builder.append(convertExpressionToMapKeyColumn(filter.getLhs()));
+          builder.append(convertStringToMapKeyColumn(getLogicalColumnName(filter.getLhs())));
           builder.append(" = ");
           builder.append(convertLiteralToString(kvp[MAP_KEY_INDEX], paramsBuilder));
           break;
         case CONTAINS_KEYVALUE:
           kvp = convertExpressionToMapLiterals(filter.getRhs());
-          String keyCol = convertExpressionToMapKeyColumn(filter.getLhs());
-          String valCol = convertExpressionToMapValueColumn(filter.getLhs());
-          builder.append(keyCol);
-          builder.append(" = ");
-          builder.append(convertLiteralToString(kvp[MAP_KEY_INDEX], paramsBuilder));
-          builder.append(" AND ");
-          builder.append(valCol);
-          builder.append(" = ");
-          builder.append(convertLiteralToString(kvp[MAP_VALUE_INDEX], paramsBuilder));
-          builder.append(" AND ");
-          builder.append(MAP_VALUE);
-          builder.append("(");
-          builder.append(keyCol);
-          builder.append(",");
-          builder.append(convertLiteralToString(kvp[MAP_KEY_INDEX], paramsBuilder));
-          builder.append(",");
-          builder.append(valCol);
-          builder.append(") = ");
-          builder.append(convertLiteralToString(kvp[MAP_VALUE_INDEX], paramsBuilder));
+          String keyCol = convertStringToMapKeyColumn(getLogicalColumnName(filter.getLhs()));
+          String valCol = convertStringToMapValueColumn(getLogicalColumnName(filter.getLhs()));
+          builder.append(getStringForMapAttribute(kvp, keyCol, valCol, paramsBuilder));
           break;
         default:
-          rhs = handleValueConversionForLiteralExpression(filter.getLhs(), filter.getRhs());
-          builder.append(
-              convertExpression2String(filter.getLhs(), paramsBuilder, executionContext));
-          builder.append(" ");
-          builder.append(operator);
-          builder.append(" ");
-          builder.append(convertExpression2String(rhs, paramsBuilder, executionContext));
+          if (filter.getLhs().getValueCase() == ATTRIBUTE_EXPRESSION) {
```

Comment:
If we need to handle this across each path, we'll run into bugs (does this support the like case above, for example?). Can we have one spot where we just do a general conversion of expressions that all paths (filter, select, group, order) go through?

Reply:
This PR was just for `CONTAINS_KEYVALUE`. Other things would be added in the next PR (Ref : https://github.com/hypertrace/query-service/pull/97)

Reply:
Discussed this a bit offline, but this is what I meant in https://github.com/hypertrace/query-service/pull/96#discussion_r750649816 . If we never add attribute expression support for CONTAINS_KEY and CONTAINS_KEYVALUE, that's fine - those are meant to be replaced by attribute expressions anyway. What we need to do in this PR (or in multiple) is everywhere we resolve an `Expression` into a logical column name add support for that `Expression` to either be of type `ColumnIdentifier` (e.g. `TRACE.id -> id`) or of type `AttributeExpression`. AttributeExpression could be a simple translation like `TRACE.id -> id`, or a complex translation like `SPAN.tags w/ subpath span.kind -> mapValue(tags__KEYS, 'span.kind' ,tags__VALUES)`

This is what gives us full support for referencing custom keys across all operations

Reply:
We also don't want to do this ad-hoc in each spot, so this should happen inside `convertExpression2String`, or something like it (who names with numbers!?)